### ISSUE TITLE
LOOKUP not supported in v2 engine

### ISSUE THREAD

**Issue by egalpin** [apache/pinot#11651]
This may be an intentional choice but LOOKUP (dimension table use) does not appear to work in v2 engine. Repro steps are:

- write any query using LOOKUP in v1 engine, validate that it works
- issue the same query to the v2 engine
- note that it fails with a message like:
```
Caused by: org.apache.calcite.runtime.CalciteContextException: From line 2, column 63 to line 2, column 126: No match found for function signature LOOKUP(<CHARACTER>, <CHARACTER>, <CHARACTER>, <NUMERIC>)
	at jdk.internal.reflect.GeneratedConstructorAccessor219.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.calcite.runtime.Resources$ExInstWithCause.ex(Resources.java:505)
...
Caused by: org.apache.calcite.sql.validate.SqlValidatorException: No match found for function signature LOOKUP(<CHARACTER>, <CHARACTER>, <CHARACTER>, <NUMERIC>)
	at jdk.internal.reflect.GeneratedConstructorAccessor218.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.apache.calcite.runtime.Resources$ExInstWithCause.ex(Resources.java:505))
```

**Comment by Jackie-Jiang** [comment]
@walterddr Do you think we need `LOOKUP` in multi-stage engine? For completeness, we might want to just annotate it so that it can be used in multi-stage engine

**Comment by walterddr** [comment]
let me dig deeper into the problem. IMO there's 2 aspect of this issue: 
1. does multi-stage engine support pre-distributed dim-tables --> i need to double check and see the dispatch logic to handle pre-distributed table
2. does multi-stage engine support LOOKUP syntax --> the answer is no because multi-stage engine supports standard SQL way to express this:
```
SELECT localCol, LOOKUP('foreignTbl', 'foreignCol', 'foreignKey', localKey) FROM tbl
```
is equivalent to
```
SELECT localCol, foreignCol
FROM tbl LEFT JOIN foreignTbl ON localKey = foreignKey
```
3. should we provide an upgrade / backward-compatible transition --> i think it is yes but this, along with SUBQUERY/IDSET and other V1 only functionalities, should be handle systematically. 
several ideas i can think of:
    a. a standalone query rewriter that translates old syntax to new
    b. automatic fallback to v1 syntax (default with v2) until the next release 
    c. automatic fall-forward to v2 syntax (default with v1) until the next release

**Comment by egalpin** [comment]
It's perhaps worth noting that LOOKUP yielded significantly faster results in some cases (depends on the query I'm sure), so if it's possible to keep it as something compatible with v2 engine that would be great to have.

**Comment by walterddr** [comment]
good point. i think that echos back to the 1st point i mentioned above (e.g. the table is recognized by v2, but it is not considered as pre-distributed table :-( and thus data shuffling will still occur) 

@Jackie-Jiang does force the engine think the table is co-partitioned help here? dim-table doesn't have a partition key so i am not sure if this is even possible, we might need a "pre-distributed table abstraction"

**Comment by walterddr** [comment]
it is very likely possible but it will only work with v1 and will not work in many of the v2 features. 

IMO, we can definitely go down this path by registering an OPAQUE function in v2 and only allow it to run on certain circumstances (similar to ARRAY_TO_MV special logic that only works in v1). 

However we should throw explicit warnings and indicate this is going to be deprecated - with the recent warning UI changes: https://github.com/apache/pinot/pull/11611 I think it should be relatively easy for users to comprehend

(assigning to myself)

**Comment by Jackie-Jiang** [comment]
@walterddr Will it throw exception if we pass ANY type into an aggregation such as `SUM`? Or it simply skips the casting?

**Comment by walterddr** [comment]
it will likely not work. let me add tests cases to the PR to demonstrate

**Comment by walterddr** [comment]
based on the draft limitation: https://github.com/apache/pinot/pull/11686 i dont think there's much value to support LOOKUP as a function in v2. basically the limitation means we can't determine the return type unless there's a way to explicitly parse that. 

i will propose a different approach to LOOKUP join, one that doesn't require user to understand this Pinot specific dialect

**Comment by walterddr** [comment]
here is a proposal for creating lookup join support in the v2 engine: https://docs.google.com/document/d/1cyFrW21MlcbHEEfl46paKRs6v6qHzwOZbIsnFL7UVnw/edit#heading=h.xdbe7h7na0kx

TL;DR the syntax could look as the following
```
SELECT /*+ joinOptions(join_strategy = ‘lookup’) */
  b.playerName, b.teamID, d.teamName
FROM baseballStats as b LEFT JOIN dimBaseballTeams as d 
  ON b.teamID = d.teamID
```

**Comment by cbalci** [comment]
Thanks for the investigation and the proposal @walterddr. 'LOOKUP' as a join strategy makes more sense to me.

I also imagine we could automatically set the join_strategy to 'lookup', if the RHS is a dim table. This would eliminate the need for the hint. What do you think?

**Comment by walterddr** [comment]
automating hints are a much wider topic (getting cost-factory to hook-in, getting table metadata to feed into cost factory etc) and i would like to keep that separate.

but to answer your question, auto setting join to lookup if RHS is dim table doesn't work --> 
- there's no guarantee the left side is a normal table, or a join result, or a temp table, etc.
- there's also situation that LHS is a semi-join result following by a LOOKUP, which can apply automatically but can't be determine until we have cost-factory and table metadata feed in.
